polygeist/lib/Dialect/Polygeist/Transforms/DetectReduction.cpp

```diff
@@ -519,7 +605,12 @@ void DetectReductionPass::runOnOperation() {
   RPS.add<AffineForReductionIter, SCFForReductionIter>(ctx, numReductions,
                                                        aliasAnalysis);
   GreedyRewriteConfig Config;
-  (void)applyPatternsAndFoldGreedily(getOperation(), std::move(RPS), Config);
+  SmallVector<Operation *> LoopOps;
+  getOperation()->walk([&](Operation *Op) {
+    if (isa<LoopLikeOpInterface>(Op))
+      LoopOps.push_back(Op);
+  });
+  (void)applyOpPatternsAndFold(LoopOps, std::move(RPS), Config);
```

Comment:
To prevent the loop in the else block to keep versioning, we generate a list of loop operations first, and only optimize those.